### PR TITLE
Validate input when setting the QuantumCircuit data attribute directly

### LINKED ISSUE #2826: Deprecate writing to QuantumCircuit.data
Similar to #2170 , but for `QuantumCircuit`. The possibility of arbitrary data in `circuit.data` makes modifying or replacing the internals of `QuantumCircuit` difficult and error prone (see discussion around `circuit._append` in #2824 ). We should:
- Move `circuit.data` to `circuit._data`
- Add a getter for `circuit.data` and document it as a ordered list of `(Instruction, qargs, cargs)`. It's an open question whether or not we should commit to preserving user-insertion order.
- Add a setter for `circuit.data` with a `DeprecationWarning`, directing users to e.g. `QuantumCircuit.append` and the rest of the circuit API.

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit adjusts how users can set a QuantumCircuit's data attribute
directly. Accessing and reading it has no changes, but when just directly
changing it's value outside of the circuits API we now will validate the
input to ensure that the value being used conforms to the expectations
and data format used throughout the circuit's api. Prior to this patch it
was possible to assign arbitrary data in the data parameter which would
make it difficult to add performant modifiers and methods in the circuit's
api because we would constantly have to verify the state of the data
attribute. This commit should fix this because now that the data is
always validated when we build out the circuit's api in the future.

Co-Authored-By: Kevin Krsulich <kevin.krsulich@ibm.com>

### Details and comments

Fixes #2826

### CODE REVIEW

**Comment by kdk** [comment]
Added a few commits to make `QuantumCircuit.data` a list-like `MutableSequence` wrapper, with the goal being that existing operations on `circuit.data` should continue to work, while maintaining a consistent state in `circuit._data`.

**Comment by mtreinish** [comment]
It looks like tests are failing because we have to update the pickle files in the mapper tests. I hit this before when I first added the `_data` attribute. (I'm still not sure why we need to save objects as a pickle instead of just generating things in a setup or setupclass method)

File: qiskit/circuit/quantumcircuitdata.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017.
```
Comment (1ucian0) [inline qiskit/circuit/quantumcircuitdata.py]:
```suggestion
# (C) Copyright IBM 2019.
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -15,6 +15,12 @@ The format is based on [Keep a Changelog].
 
 ## [UNRELEASED]
 
+### Changed
+
+- Assignments and modifications to `QuantumCircuit.data` will now be broadcast
+  and validated following the same rules used throughout the
+  QuantumCircuit` API. (\#2826)
```
Comment (1ucian0) [inline CHANGELOG.md]:
```suggestion
  `QuantumCircuit` API. (\#2826)
```

**Pull request merged**